File: implementations/plutarch/lib/UplcBenchmark/PoolValidator.hs
module UplcBenchmark.PoolValidator (ppoolValidator) where

import Data.Kind (Type)
import GHC.Generics qualified as GHC
import Generics.SOP qualified as SOP
import Plutarch.LedgerApi.V3 (
  PMaybeData (PDJust),
  POutputDatum (POutputDatum),
  PScriptContext (PScriptContext),
  PScriptInfo (PSpendingScript),
  PTxInInfo (PTxInInfo),
  PTxInfo (PTxInfo),
  PTxOut (PTxOut),
 )
import Plutarch.LedgerApi.Value (PAssetClass (PAssetClass), PCurrencySymbol, pvalueOf)
import Plutarch.Monadic qualified as P
import Plutarch.Unsafe (punsafeCoerce)
import UplcBenchmark.Utils (
  passert,
  pisPDNothing,
  ptryFindInput,
  ptryFindOutputWithAsset,
  punsafeCastDatum,
  punsafeCastRedeemer,
  pvalueOfAssetClass,
 )

type DexDatum :: S -> Type
data DexDatum s = DexDatum
  { tokenA :: Term s (PAsData PAssetClass)
  , tokenB :: Term s (PAsData PAssetClass)
  , poolNft :: Term s (PAsData PAssetClass)
  , lpToken :: Term s (PAsData PCurrencySymbol)
  , mintedLpTokens :: Term s (PAsData PInteger)
  , swapFee :: Term s (PAsData PInteger)
  }
  deriving stock (GHC.Generic)
  deriving anyclass (SOP.Generic, PEq, PIsData)
  deriving (PlutusType) via (DeriveAsDataRec DexDatum)

type DexRedeemer :: S -> Type
data DexRedeemer s
  = DexRedeemer'Swap
  | DexRedeemer'DepositLiquidity
  | DexRedeemer'WithdrawLiquidity
  deriving stock (GHC.Generic, Show)
  deriving anyclass (SOP.Generic, PEq, PIsData)
  deriving (PlutusType) via DeriveTagPlutusType DexRedeemer

-- TODO: PTryFrom

ppoolValidator :: ClosedTerm (PAsData PScriptContext :--> POpaque)
ppoolValidator = plam $ \ctx' -> P.do
  PScriptContext txInfo redeemer info <- pmatch (pfromData ctx')
  PTxInfo inputs _ outputs _ mint _ _ _ _ _ _ _ _ _ _ _ <- pmatch txInfo
  PSpendingScript ownTxOutRef datum' <- pmatch info
  PDJust datum <- pmatch datum'
  DexDatum inTokenA inTokenB inPoolNft' inLpToken inMintedLpTokens inSwapFee <- pmatch (pfromData $ punsafeCastDatum $ pfromData datum)
  inPoolNft <- plet $ pfromData inPoolNft'
  nftAndLpTokenName <- plet $ pmatch inPoolNft $ \(PAssetClass inPoolNft'') -> psndBuiltin # inPoolNft''

  (PTxInInfo _ ownInputResolved) <- pmatch $ pfromData (ptryFindInput # ownTxOutRef # pfromData inputs)
  (PTxOut _ ownInputValue _ _) <- pmatch ownInputResolved

  (PTxOut _ ownOutputValue ownOutputDatum ownOutputReferenceScript) <- pmatch $ pfromData (ptryFindOutputWithAsset # inPoolNft # pfromData outputs)

  let hasNft = plam $ \value -> pvalueOfAssetClass # value # inPoolNft #== 1

  newMintedLp <- plet $ pvalueOf # pfromData mint # pfromData inLpToken # pfromData nftAndLpTokenName

  inAAmount <- plet $ pvalueOfAssetClass # pfromData ownInputValue # pfromData inTokenA
  inBAmount <- plet $ pvalueOfAssetClass # pfromData ownInputValue # pfromData inTokenB

  outAAmount <- plet $ pvalueOfAssetClass # pfromData ownOutputValue # pfromData inTokenA
  outBAmount <- plet $ pvalueOfAssetClass # pfromData ownOutputValue # pfromData inTokenB

  POutputDatum outputDatum' <- pmatch ownOutputDatum
  outputDatum <- plet $ pfromData $ punsafeCoerce @(PAsData DexDatum) @PData $ pto outputDatum'
  DexDatum outTokenA outTokenB outPoolNft outLpToken outMintedLpTokens outSwapFee <- pmatch outputDatum

  passert "Input UTxO must have NFT" (hasNft # pfromData ownInputValue)

  passert "Output cannot have ref script" (pisPDNothing # ownOutputReferenceScript)

  passert "Invalid output datum: tokenA" (inTokenA #== outTokenA)
  passert "Invalid output datum: tokenB" (inTokenB #== outTokenB)
  passert "Invalid output datum: poolNft" (pdata inPoolNft #== outPoolNft)
  passert "Invalid output datum: lpToken" (inLpToken #== outLpToken)
  passert "Invalid output datum: swapFee" (inSwapFee #== outSwapFee)

  pmatch (pfromData $ punsafeCastRedeemer redeemer) $ \case
    -- NOTE: Can we just make these two checks be the same?
    DexRedeemer'DepositLiquidity -> P.do
      passert
        "Invalid output datum: mintedLpTokens"
        (pfromData inMintedLpTokens + newMintedLp #== pfromData outMintedLpTokens)

      passert "Must not burn LP tokens" (newMintedLp #> 0)

      passert
        "Not deposited enough"
        (pfromData outMintedLpTokens * pfromData outMintedLpTokens #<= outAAmount * outBAmount)

      popaque $ pconstant @PUnit ()
    DexRedeemer'WithdrawLiquidity -> P.do
      passert
        "Invalid output datum: mintedLpTokens"
        (pfromData inMintedLpTokens + newMintedLp #== pfromData outMintedLpTokens)

      passert "Must not mint LP tokens" (newMintedLp #< 0)

      passert
        "Withdrawn too much"
        (pfromData outMintedLpTokens * pfromData outMintedLpTokens #<= outAAmount * outBAmount)

      popaque $ pconstant @PUnit ()
    DexRedeemer'Swap -> P.do
      passert
        "Invalid output datum: mintedLpTokens"
        (inMintedLpTokens #== outMintedLpTokens)

      passert "Must not mint LP tokens" (newMintedLp #== 0)

      passert
        "Swapped too much"
        (pcheckSwap # pfromData inSwapFee # inAAmount # inBAmount # outAAmount # outBAmount)

      popaque $ pconstant @PUnit ()

pcheckSwap ::
  Term
    s
    ( PInteger
        :--> PInteger
        :--> PInteger
        :--> PInteger
        :--> PInteger
        :--> PBool
    )
pcheckSwap = plam $ \feeNum oldA oldB newA newB ->
  feeDen * feeDen * oldA * oldB #<= ((newA * feeDen) - ((getDiff # newA # oldA) * feeNum)) * ((newB * feeDen) - ((getDiff # newB # oldB) * feeNum))

-- The uniswap fee is 0.3%; here it is multiplied by 1000, so that the
-- on-chain code deals only in integers.
-- See: <https://uniswap.org/whitepaper.pdf> Eq (11) (Page 7.)
feeDen :: ClosedTerm PInteger
feeDen = 1000

getDiff :: Term s (PInteger :--> PInteger :--> PInteger)
getDiff = phoistAcyclic $
  plam $ \new old ->
    pif (new - old #<= 0) 0 (new - old)

File: implementations/opshin/src/pool_validator.py
#!opshin
from opshin.prelude import *


@dataclass()
class AssetClass(PlutusData):
    CONSTR_ID = 0
    symbol: PolicyId
    name: TokenName


@dataclass()
class DexDatum(PlutusData):
    token_a: AssetClass
    token_b: AssetClass
    pool_nft: AssetClass
    lp_token: PolicyId
    minted_lp_tokens: int
    swap_fee: int


def decode_dex_datum(raw: List[Anything]) -> DexDatum:
    token_a: AssetClass = raw[0]
    token_b: AssetClass = raw[1]
    pool_nft: AssetClass = raw[2]
    lp_token: PolicyId = raw[3]
    minted_lp_tokens: int = raw[4]
    swap_fee: int = raw[5]
    return DexDatum(token_a, token_b, pool_nft, lp_token, minted_lp_tokens, swap_fee)


DexRedeemerSwap: int = 0
DexRedeemerDepositLiquidity: int = 1
DexRedeemerWithdrawLiquidity: int = 2


def quantity_of_asset(value: Value, asset_class: AssetClass) -> int:
    return value[asset_class.symbol][asset_class.name]


def quantity_of(value: Value, symbol: PolicyId, name: TokenName) -> int:
    if symbol in value.keys():
        return value[symbol].get(name, 0)
    else:
        return 0


def get_diff(new: int, old: int) -> int:
    if new - old < 0:
        return 0
    else:
        return new - old


# The uniswap fee is 0.3%; here it is multiplied by 1000, so that the
# on-chain code deals only in integers.
# See: <https://uniswap.org/whitepaper.pdf> Eq (11) (Page 7.)
fee_den: int = 1000


def check_swap(
    fee_num: int,
    old_a: int,
    old_b: int,
    new_a: int,
    new_b: int,
) -> bool:
    return fee_den * fee_den * old_a * old_b <= (
        new_a * fee_den - get_diff(new_a, old_a) * fee_num
    ) * (new_b * fee_den - get_diff(new_b, old_b) * fee_num)


def validator(
    input_datum_raw: List[Anything], redeemer: int, context: ScriptContext
) -> None:
    input_datum: DexDatum = decode_dex_datum(input_datum_raw)

    purpose = context.purpose
    if isinstance(purpose, Spending):
        own_input_ref = purpose.tx_out_ref
    else:
        assert False, "Invalid purpose"

    own_input = [inp for inp in context.tx_info.inputs if inp.out_ref == own_input_ref][
        0
    ]

    own_output = [
        out
        for out in context.tx_info.outputs
        if quantity_of_asset(out.value, input_datum.pool_nft) == 1
    ][0]

    # lp name == nft name
    new_minted_lp = quantity_of(
        context.tx_info.mint, input_datum.lp_token, input_datum.pool_nft.name
    )

    in_a_amount = quantity_of_asset(own_input.resolved.value, input_datum.token_a)
    in_b_amount = quantity_of_asset(own_input.resolved.value, input_datum.token_b)

    out_a_amount = quantity_of_asset(own_output.value, input_datum.token_a)
    out_b_amount = quantity_of_asset(own_output.value, input_datum.token_b)

    out_datum_some = own_output.datum
    if isinstance(out_datum_some, SomeOutputDatum):
        out_datum_raw: List[Anything] = out_datum_some.datum
    else:
        assert False, "Missing inline output datum"
    output_datum = decode_dex_datum(out_datum_raw)

    assert quantity_of_asset(own_input.resolved.value, input_datum.pool_nft) == 1
    own_output_reference_script = own_output.reference_script
    assert isinstance(own_output_reference_script, NoScriptHash)
    assert input_datum.token_a == output_datum.token_a
    assert input_datum.token_b == output_datum.token_b
    assert input_datum.pool_nft == output_datum.pool_nft
    assert input_datum.lp_token == output_datum.lp_token
    assert input_datum.swap_fee == output_datum.swap_fee

    if redeemer == DexRedeemerDepositLiquidity:
        assert (
            input_datum.minted_lp_tokens + new_minted_lp
            == output_datum.minted_lp_tokens
        )
        assert new_minted_lp > 0
        assert (
            output_datum.minted_lp_tokens * output_datum.minted_lp_tokens
            <= out_a_amount * out_b_amount
        )
    elif redeemer == DexRedeemerWithdrawLiquidity:
        assert (
            input_datum.minted_lp_tokens + new_minted_lp
            == output_datum.minted_lp_tokens
        )
        assert new_minted_lp < 0
        assert (
            output_datum.minted_lp_tokens * output_datum.minted_lp_tokens
            <= out_a_amount * out_b_amount
        )
    elif redeemer == DexRedeemerSwap:
        assert input_datum.minted_lp_tokens == output_datum.minted_lp_tokens
        assert new_minted_lp == 0
        assert check_swap(
            input_datum.swap_fee,
            in_a_amount,
            in_b_amount,
            out_a_amount,
            out_b_amount,
        )
    else:
        assert False, "Invalid redeemer"
